### PR TITLE
Feature branch

### PR DESCRIPTION
**Pull request by george-shammar**
This Pull Request seeks to merge feature_branch to the development branch with the following changes:

The following ruby enumerable methods were created in the file.
1. my_each 
2. my_each_with_index
3. my_select
4. my_all
5. my_any
6. my_none
7. my_count
8. my_map
9. my_inject

my_inject method was tested by creating a method called multiply_els.

my_map method was modified to accept proc instead of block.

### CODE REVIEW

**Comment by descholar-ceo** [comment]
# This Pull Request is under review by [descholar-ceo](https://github.com/descholar-ceo) :eyes: :eyes: 
## Please wait a bit ...

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,78 @@
+module Enumerable
+  def my_each(arr = to_a)
+    arr.length.times do |object|
+      yield arr[object]
+    end
+  end
```
Comment (descholar-ceo) [inline enumerables.rb:6]: ## `#my_each` should have the following behaviors:
- [ ] returns an Enumerator if no block is given
- [ ] when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- [ ] when a block is given when self is a range returns the range itself after calling the given block once for each element in self

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+  def my_each_with_index
+    arr = to_a
+    arr.length.times do |object|
+      yield arr[object], object
+    end
+  end
```
Comment (descholar-ceo) [inline enumerables.rb:13]: ## `#my_each_with_index` should have the following behaviors:
- [ ] returns an enumerator if no block is given
- [ ] when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- [ ] when a block is given when self is a range returns the range itself after calling the given block once for each element in self

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+  def my_select
+    array = []
+    my_each do |object|
+      array.push(object) if yield(object)
+    end
+    array
+  end
```
Comment (descholar-ceo) [inline enumerables.rb:21]: ## `#my_select` should have the following behaviors:
- [ ] returns an enumerator if no block is given

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+  def my_all?
+    my_each do |object|
+      return true if yield(object) == true
+
+      return false
+    end
+  end
```
Comment (descholar-ceo) [inline enumerables.rb:29]: ## `#my_all` should have the following behaviors:
- [ ] when no block or argument is given returns true when none of the collection members are false or nil
- [ ] when no block or argument is given returns false when one of the collection members are false or nil
- [ ] when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
- [ ] when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
- [ ] when a class is passed as an argument returns false if any of the collection is not a member of such class
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [ ] when a Regex is passed as an argument returns false if any of the collection does not match the Regex
- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- [ ] when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+  def my_any?
+    my_each do |i|
+      return true if yield(i) == true
+    end
+    false
+  end
```
Comment (descholar-ceo) [inline enumerables.rb:36]: ## `#my_any` should have the following behaviors:
- [ ] when no block or argument is given returns true if at least one of the collection is not false or nil
- [ ] when no block or argument is given returns false if at least one of the collection is not true
- [ ] when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
- [ ] when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
- [ ] when a Regex is passed as an argument returns true if any of the collection matches the Regex
- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
- [ ] when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+  def my_none?
+    my_each do |object|
+      return true if yield(object) == false
+
+      return false
+    end
+  end
```
Comment (descholar-ceo) [inline enumerables.rb:44]: ## `#my_none` should have the following behaviors:
- [ ] when no block or argument is given returns true only if none of the collection members is true
- [ ] when no block or argument is given returns false only if one of the collection members is true
- [ ] when a class is passed as an argument returns true if none of the collection is a member of such class
- [ ] when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
- [ ] when a class is passed as an argument returns false if any of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns true if none of the collection matches the Regex
- [ ] when a Regex is passed as an argument returns false if any of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
- [ ] hen a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

File: enumerables.rb [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+  def my_count
+    number = 0
+    my_each do |item|
+      number += 1 if yield(item) == true
+    end
+    number
+  end
```
Comment (descholar-ceo) [inline enumerables.rb:52]: ## `#my_count` should have the following behaviors:
- [ ] returns the number of items in enum through enumeration
- [ ] returns the number of items in enum through enumeration::range
- [ ] counts the number of items in enum that are equal to item if an argument is given

File: enumerables.rb [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+  def my_map(*)
+    array = []
+    my_each do |i|
+      array.push(yield(i))
+    end
+    array
+  end
```
Comment (descholar-ceo) [inline enumerables.rb:60]: ## `#my_map` should have the following behaviors:
- [ ] returns an Enumerator if no block is given
- [ ] executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

File: enumerables.rb [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+  def my_inject(accum = [0])
+    my_each do |item|
+      accum = yield(accum, item)
+    end
+    accum
+  end
```
Comment (descholar-ceo) [inline enumerables.rb:67]: ## `#my_inject` should have the following behaviors:
- [ ] does not mutate the original array
- [ ] when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block

File: enumerables.rb [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+  def multiply_els(arr)
+    arr.my_inject(1) do |accum, item|
+      accum * item
+    end
+  end
+end
+
+my_proc = proc { |x| x + 7 }
+arr = [4, 16, 9]
+puts arr.my_map(&my_proc)
```
Comment (descholar-ceo) [inline enumerables.rb:78]: - [ ] Remove this puts statemnt, this is for debugging purpose, so we don't need to commit it on our repository.

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @george-shammar  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
:+1: Your enumerables pass some test cases, 
#### :warning: However we need to make few changes to mark this project approved. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [x] Add setup and usage sections to readme file
- [ ] There some test cases which your enumerables are failing, please read and learn all of inline comments to understand more what you need to do.
___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

File: enumerables.rb [hunk 11]
```diff
@@ -0,0 +1,120 @@
+# rubocop:disable Metrics/CyclomaticComplexity
+# rubocop:disable Metrics/PerceivedComplexity
+module Enumerable
+  def my_each(arr = to_a)
+    return enum_for: my_each unless block_given?
```
Comment (ebeagusamuel) [inline enumerables.rb:5]: - [ ] You have a syntax error here. Update it to `enum_for(:my_each)`. Check all the methods where you have this and update it there too.

**Review by ebeagusamuel — CHANGES_REQUESTED** [review]
# Changes Required :recycle: 
### Hey there :wave: 
#### Nice work so far, but there are still changes to be done as some of the test cases are still failing.
- [x] Please format the setup section in your readme to be more readable and understandable.
- [x] You have syntax error causing an error when I want to run the program. Check my inline comment.
- [x] I will list all the cases that are still failing.
-  #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
- #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
- #my_all? when a class is passed as an argument returns false if any of the collection is not a member of such class
- #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
- #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
- #my_count returns the number of items in enum through enumeration
- #my_count counts the number of items in enum that are equal to item if an argument is given
-  #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
- #my_inject searches for the longest word in an array of strings
-  #my_inject does not mutate the original array
-  #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block
- #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range
- #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
- #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
- #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method
- #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range
- #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject
### Request another review when you made the changes 
slack(@ebeagusamuel)

**Comment by mimipeshy** [comment]
## STATUS :CHANGES REQUIRED :red_circle: :recycle: 

Hello @george-shammar :wave: :wave: good job on working on all the feedback given by the TSE's, this is impressive :clap: :clap: However, let's fix two things before marking the project as complete

- [ ]  NoMethodError:  undefined method `multiply_els' =>multiply_els is not an actual ruby enumerable, so move the method and place it outside the Module enumerable.
- [ ] In the README file author section, kindly add actual links to the social media links, it's easier for someone to find you that way.
That's all :smile: , Kindly submit another review when done

In case you have any questions, feel free to slack me anytime, will be glad to help you.

Regards
Peris Ndanu

**Review by Georjane — APPROVED** [review]
## Status: PR Approved 💯 🥇 
### Hi @george-shammar ✋🏽 
Congratulations on successfully completing the enumerable methods. All tests are passing now. It's a thumbs up for me. 👍🏽 You can go ahead and merge this pull request.

Good luck on your next project and Happy Coding 💻

**Pull request merged**